### ISSUE TITLE
What does normalizedParameters actually contain in makeCredential?

### ISSUE THREAD

**Issue by bzbarsky** [w3c/webauthn#265]
The spec says:

> Add a new object of type ScopedCredentialParameters to normalizedParameters, with type set to current.type and algorithm set to normalizedAlgorithm.

But "normalizedAlgorithm" is an IDL dictionary, while the `algorithm` member of a `ScopedCredentialParameters` dictionary is a `(DOMString or object)`.  Is the intention that "normalizedAlgorithm" is converted to an ES object using <https://heycam.github.io/webidl/#dictionary-to-es>?  Is the intention that normalizedParameters be a list of (type, algorithm) pairs where the algorithm is an IDL dictionary?  Something else?

**Comment by equalsJeffH** [comment]
@bzbarsky wrote:
>  Is the intention that normalizedParameters be a list of (type, algorithm) pairs where the algorithm is an IDL dictionary?

in looking at this I'm concluding that the answer to the above is "yeah, sorta that", however, the type-impedance-mismatch you note..
> But "normalizedAlgorithm" is an IDL dictionary, while the algorithm member of a ScopedCredentialParameters dictionary is a (DOMString or object).
```idl
dictionary ScopedCredentialParameters {         // from webauthn
    required ScopedCredentialType  type;
    required AlgorithmIdentifier   algorithm;
};

typedef (object or DOMString) AlgorithmIdentifier;  // from webcrypto
```
..means that we need to do some more explcit magic in order to assign `normalizedAlgorithm` to `ScopedCredentialParameters.algorithm`, yes?  

Might that magic be something like using the https://heycam.github.io/webidl/#dictionary-to-es conversion and saying something like..

>Add a new object of type `ScopedCredentialParameters` to `normalizedParameters`, with `type` set to `current.type` and `algorithm` set to the result of feeding `normalizedAlgorithm` to [dictionary-to-es](https://heycam.github.io/webidl/#dictionary-to-es).

..?

**Comment by bzbarsky** [comment]
You could do that.

However, then your normalizedParameters contains JS objects.  So you probably can't do asynchronous (possibly off-thread or out-of-process) things with it like you want to to in step 10 of https://w3c.github.io/webauthn/#makeCredential

**Comment by equalsJeffH** [comment]
Ok, so an underlying question here is there an IDL type that can be assigned an arbitrary dictionary type?  I.e., we have..
```idl
dictionary ScopedCredentialParameters {              // from webauthn
    required ScopedCredentialType  type;
    required AlgorithmIdentifier   algorithm;
};

typedef (object or DOMString) AlgorithmIdentifier;  // from webcrypto
```
..and we are trying to do this..
```
3. Let `normalizedParameters` be a new list.
4. [...]
- Let `normalizedAlgorithm` be the result of [=normalizing an algorithm=] [WebCryptoAPI], [...]
- Append a new object of type `ScopedCredentialParameters` to `normalizedParameters`, 
  with `type` set to `current.type` and `algorithm` set to `normalizedAlgorithm`.
```
..where `normalizedAlgorithm` is some IDL dictionary.  Is there some type we can use for `algorithm` (instead of `AlgorithmIdentifier`) such that we can do something like this..
```
3. Let `normalizedParameters` be a new list.
4. [...]
- Let `normalizedAlgorithm` be the result of [=normalizing an algorithm=] [WebCryptoAPI], [...]
- Append a new object of type `ScopedCredentialParameters` to `normalizedParameters`, 
  with `type` set to `current.type` and `algorithm` set to  the result of converting the 
  ECMAScript object represented by `normalizedAlgorithm`, to a WebIDL dictionary type, as 
  defined by [WebIDL].
```
..or does `AlgorithmIdentifier` already work because it can be of type `object`?  

It seems that the WebCrypto `normalizing an algorithm` algorithm performs similar magic and I'm wondering if we can use it as a model?

**Comment by bzbarsky** [comment]
I think just having a https://infra.spec.whatwg.org/#list that holds the things you want is reasonable.  Nothing forces the things in a "list" in that sense to be describable by IDL, even, much less a single type or whatever.

**Comment by equalsJeffH** [comment]
thx, hm, then can we do something like this?

3. Let `normalizedParameters` be a new [=list=] whose [=items=] are pairs of 
   `ScopedCredentialType` and a [=dictionary=] type (as returned by [=normalizing an algorithm=] 
   [WebCryptoAPI]).
4. [...]
- Let `normalizedAlgorithm` be the result of [=normalizing an algorithm=]   [...]
- Append a new [=item=], consisting of `current.type` and `normalizedAlgorithm`, to 
   `normalizedParameters`.

**Comment by bzbarsky** [comment]
Yes, something like that should work.  There's some concerns around the way buffer sources are handled in webcrypto there, but it should be clear enough in practice....

**Comment by equalsJeffH** [comment]
this was fixed in PR #347
